### ISSUE TITLE
Consider moving no-std/alloc error stuff from `internals` to a new stable crate

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3322]
Context: I'm preparing a presentation about error handling in Rust and it looks like a crate containing `write_err!` and `input_string` would make sense. In a way it increases our maintenance burden but it also makes some things easier by allowing us to have `InputString` in public API. It's also so trivial that it could be stabilized ~immediately which could give us more experience with the process. (BTW I have a recent experienced: yesterday I've [stabilized my first crate](https://docs.rs/fmt2io/1.0.0/fmt2io/index.html)! Also see MSRV section for some fun.)

To be clear: this doesn't propose to move all error stuff, in particular, it won't move the `ParseIntError` which is unrelated to no-std support.

**Comment by apoelstra** [comment]
Yeah, agreed. This is a pretty common need across our ecosystem and would give us some experience stabilizing crates. And it would be nice to see some 1.0s.
